libs/plugins/modifier.escape.php

```diff
@@ -34,87 +33,18 @@ function smarty_modifier_escape($string, $esc_type = 'html', $char_set = null, $
-                return htmlspecialchars($string, ENT_QUOTES, $char_set, $double_encode);
-            } else {
-                if ($double_encode) {
-                    // php <5.2.3 - only handle double encoding
-                    return htmlspecialchars($string, ENT_QUOTES, $char_set);
-                } else {
-                    // php <5.2.3 - prevent double encoding
-                    $string = preg_replace('!&(#?\w+);!', '%%%SMARTY_START%%%\\1%%%SMARTY_END%%%', $string);
-                    $string = htmlspecialchars($string, ENT_QUOTES, $char_set);
-                    $string = str_replace(
-                        array(
-                            '%%%SMARTY_START%%%',
-                            '%%%SMARTY_END%%%'
-                        ),
-                        array(
-                            '&',
-                            ';'
-                        ),
-                        $string
-                    );
-                    return $string;
-                }
-            }
+            return htmlspecialchars($string, ENT_QUOTES, $char_set, $double_encode);
         // no break
         case 'htmlall':
             if (Smarty::$_MBSTRING) {
-                // mb_convert_encoding ignores htmlspecialchars()
-                if ($_double_encode) {
-                    // php >=5.3.2 - go native
-                    $string = htmlspecialchars($string, ENT_QUOTES, $char_set, $double_encode);
-                } else {
-                    if ($double_encode) {
-                        // php <5.2.3 - only handle double encoding
-                        $string = htmlspecialchars($string, ENT_QUOTES, $char_set);
-                    } else {
-                        // php <5.2.3 - prevent double encoding
-                        $string = preg_replace('!&(#?\w+);!', '%%%SMARTY_START%%%\\1%%%SMARTY_END%%%', $string);
-                        $string = htmlspecialchars($string, ENT_QUOTES, $char_set);
-                        $string =
-                            str_replace(
-                                array(
-                                    '%%%SMARTY_START%%%',
-                                    '%%%SMARTY_END%%%'
-                                ),
-                                array(
-                                    '&',
-                                    ';'
-                                ),
-                                $string
-                            );
-                        return $string;
-                    }
-                }
+                $string = htmlspecialchars($string, ENT_QUOTES, $char_set, $double_encode);
                 // htmlentities() won't convert everything, so use mb_convert_encoding
-                return mb_convert_encoding($string, 'HTML-ENTITIES', $char_set);
+                $string = mb_convert_encoding($string, 'UTF-8', $char_set);
+                $string = htmlentities($string);
+                return htmlspecialchars_decode($string);
```

Comment:
Why the `htmlspecialchars_decode`? This will cause characters like <, > " and ? to not be escaped.

Reply:
Wouldn't this trigger errors in `PluginModifierEscapeTest`?

Reply:
Tried, with ou without the l.44 and no error in test.

Reply:
Is there a bug in PHPUnit Tests ?

Reply:
You are right. Although, interestingly, `smarty_modifier_escape` is not covered in `PluginModifierEscapeTest`. It actually calls `smarty_modifiercompiler_escape`. `smarty_modifier_escape` is only used when running in nocache mode. Still, I'd consider it useful to make the implementations equivalent. Now, the order of htmlspecialchars_decode, mb_convert_encoding, htmlentities and htmlspecialchars differs from `smarty_modifiercompiler_escape`. Both are so insanely complex that I'm not sure which to pick :smile:

Reply:
May be to make in an another PR. :smile_cat:

Reply:
Actually, `smarty_modifier_escape` is used when one of the parameters is not a literal and is thus not "evaluatable" at compile time.

Reply:
Thanks for the last commit. Do you need anything else in this PR ?

Reply:
@wisskid Thanks for the last commit. Do you need anything else in this PR ?